### PR TITLE
Hyperparameter tuning, FocalDiceLoss, and 5M/10M cross-regime transfer evaluation

### PR DESCRIPTION
**Pull request by sridhs21**
## Summary
This PR changes: 1 file modified (`XPointMLTest.py`), 3 new files (`optuna_tuner.py`, `test_xpoint_transfer.py`, `build_transfer_cache.py`).

## What was done:

- **Optuna-driven hyperparameter tuning** (`optuna_tuner.py`) — TPE sampler + median pruner over base_channels, dropout, weight decay, learning rate, positive-patch ratio, focal/dice weighting, scheduler choice, and SWA start fraction. Replaces the prior ad-hoc grid where `base_channels` was hard-coded to 32 while tuners assumed 64.

- **FocalDiceLoss + LR scheduling + SWA** — new loss combining focal cross-entropy and Dice with configurable α / γ / dice weight; linear LR warmup followed by cosine annealing (or ReduceLROnPlateau); optional Stochastic Weight Averaging with a custom BN-update step compatible with the dict-based dataloader.

- **Cross-regime transfer evaluation** (`test_xpoint_transfer.py`) — loads the best PKPM-trained checkpoint and evaluates zero-shot on 5M and 10M Gkeyll datasets (150 frames each), producing per-dataset and combined summaries. Re-evaluates the PKPM validation set as an in-domain reference.

- **Cache build pipeline** (`build_transfer_cache.py`) — precomputes the deterministic X-point finder for all 150 frames of each transfer dataset, so subsequent evaluation runs read `.npy` caches instead of re-parsing `.gkyl` files. Supports `--workers N` for parallel processing and `RC_EXTRACT_DIR` / `RC_CACHE_BASE` env-var path overrides for ramdisk staging.

- **Augmentation correctness fixes** in `XPointMLTest.py` — brightness/contrast jitter is now applied globally (not per channel) so the physical identities `Bx = ∂y ψ`, `By = -∂x ψ`, `Jz = -∇²ψ/μ₀` stay consistent across the four input channels; cutout no longer mutates cached frame tensors in place.

- **Profiling and minor perf in `getPgkylData`** — per-stage `[PROFILE]` timings around `compactRead`, gradient computation, `getCritPoints`, and `getXOPoints`; Hessian is now packed from already-computed second derivatives and passed to `getXOPoints(hessian=…)` to avoid recomputing gradients.

### CODE REVIEW

File: optuna_tuner.py [hunk 1]
```diff
@@ -0,0 +1,614 @@
+"""
+Optuna hyperparameter tuner for XPointMLTest.py
+
+Usage:
+    python optuna_tuner.py \
+        --paramFile /path/to/params.txt \
+        --xptCacheDir /path/to/cache \
+        --n-trials 50 \
+        --study-name xpoint-tuning \
+        --db sqlite:///optuna_xpoint.db
```
Comment (cwsmith) [inline optuna_tuner.py:10]: does optuna automatically create the db or are additional manual setup steps required?

File: test_xpoint_transfer.py [hunk 2]
```diff
@@ -0,0 +1,273 @@
+#!/usr/bin/env python3
+"""
+Cross-domain inference: evaluate the best PKPM-trained model on 5M and 10M data.
+
+This script:
+  1. Extracts 5M.tgz and 10M.tgz (if not already extracted)
```
Comment (cwsmith) [inline test_xpoint_transfer.py:6]: We should force this to use the cache if `XPointMLTest.py` requires it.

File: XPointMLTest.py
Comment (cwsmith) [inline XPointMLTest.py:968]: IIRC, this option will run the hessian based classifier and build the cache.  How does this differ from the new `build_transfer_cache.py`?  If they do the same thing we should likely remove the option, and supporting functionality, here and require the use of the cache prepared with `build_transfer_cache.py`.

On that note, we should probably rename `build_transfer_cache.py` to `run_hessian_and_build_cache.py` or something similarly explicit.

Comment (cwsmith) [inline XPointMLTest.py]: this looks like the call to run the hessian based classifier and write the cache

**Review by cwsmith — COMMENTED** [review]
Thank you.  A few comments are below.

**Comment by cwsmith** [comment]
IIRC, a patch was needed for an indexing bug in https://github.com/SCOREC/pgkylFrontEnd.  If so, would you please create a PR with that change?
